### DISCUSSION TITLE
Asserting Type Guards that works with mypy ?

### DISCUSSION THREAD

**Discussion by tushar5526** [python/typing#1527] [General]
I am not sure if this is the right place to ask this question, but I have been bugged with a simple `typing` problem of needing a behaviour where I can have something like an "asserting" `TypeGuard` that works with mypy. 

```python
def check_obj_of_type(
    obj: typing.Any, validation_type: typing.Type[T], raise_error: bool = False
) -> TypeGuard[T]:
    """
    Generic Type Guard, that checks for a type and raises Type Error
    """
    if isinstance(obj, validation_type):
        return True
    if raise_error:
        raise TypeError(f"Object should be {validation_type} not '{type(obj)}'")
    return False
```

If I want to use `check_obj_of_type` with say a function like

```py
class SomeClass:
    ...
    def do_some_stuff() -> "SomeType":
        if check_obj_of_type(self.some_var, SomeType, raise_error=True):
           return "SomeType"
    ...
```
`mypy` will error out for the above code stating `Missing return statement` but in-fact the code will never reach the `else` clause. 

For `mypy` to pass successfully, one would have to either `#type ignore[return]` or add a an assert in `else` clause.  Is there any other better way to do this ?

**Comment by erictraut** [comment]
There is currently no support in the Python type system for a "type assert". Other languages like TypeScript have this sort of feature.

There was [some discussion of adding a `TypeAssert` feature](https://github.com/python/typing/discussions/1013#discussioncomment-1966238), but there wasn't sufficient interest to move to the PEP stage with this proposal.

If you're able to modify the code in question, you can change the "check" function to return the value if its type matches and raise an error if not. 

```python
from typing import Any, Literal, TypeVar, overload

T = TypeVar("T")

@overload
def check_obj_of_type(
    obj: Any, validation_type: type[T], raise_error: Literal[True]
) -> T:
    ...

@overload
def check_obj_of_type(
    obj: Any, validation_type: type[T], raise_error: bool = False
) -> Any:
    ...

def check_obj_of_type(
    obj: Any, validation_type: type[T], raise_error: bool = False
) -> T | Any:
    if isinstance(obj, validation_type):
        return obj
    if raise_error:
        raise TypeError(f"Object should be {validation_type} not '{type(obj)}'")
    return False

class SomeType:
    ...

class SomeClass:
    some_var: Any

    def do_some_stuff(self) -> SomeType:
        return check_obj_of_type(self.some_var, SomeType, raise_error=True)
```

**Reply by tushar5526** [reply]
Thanks @erictraut I will try some workarounds!
